Preferred browser setting and open links in browser

joeyg opened the pull request:
Resolves #170
Resolves #164 

![simulator screen shot - iphone 8 - 2018-04-06 at 13 03 34](https://user-images.githubusercontent.com/490833/38441841-f3f6a46c-399a-11e8-8ab4-9547772877b7.png)
![simulator screen shot - iphone 8 - 2018-04-06 at 13 03 38](https://user-images.githubusercontent.com/490833/38441848-f5ef6696-399a-11e8-9302-06ef2414b276.png)

sashei commented on lockbox-ios/Presenter/ItemDetailPresenter.swift:
```diff
@@ -40,26 +40,35 @@ class ItemDetailPresenter {
+                        .disposed(by: target.disposeBag)
+            } else if value == Constant.string.webAddress {
+                target.dataStore.onItem(self.view?.itemId ?? "")
                     .take(1)
```
nit: indentation should match the previous conditional block

sashei commented on lockbox-ios/Presenter/PreferredBrowserSettingPresenter.swift:
```diff
@@ -0,0 +1,73 @@
+
+    func onViewReady() {
+        let driver = self.userDefaults.rx.observe(String.self, SettingKey.preferredBrowser.rawValue)
+            .map { value -> PreferredBrowserSetting? in
```
you can use `.filterNil()` to unwrap the optional here

joeyg replied:
Where should the default value come from? If there is no value in user defaults this would stop any of the options from being shown. Should default values be stored in userdefaults on first app launch?

sashei commented on lockbox-ios/Presenter/RootPresenter.swift:
```diff
@@ -199,11 +201,30 @@ class RootPresenter {
+
+    private func openUrl(string url: String) {
+        UserDefaults.standard.rx.observe(String.self, SettingKey.preferredBrowser.rawValue)
+            .take(1)
```
👍 nice

sashei commented on lockbox-ios/Store/UserInfoStore.swift:
```diff
@@ -134,3 +135,53 @@ enum AutoLockSetting: String {
+        case .Safari:
+            urlToOpen = URL(string: "http://\(url)")
+        case .Firefox:
+            urlToOpen = URL(string: "firefox://open-url?url=http://\(url)")
```
Have you tested that this works? I know there is some funny URL encoding required to pass full URLs to Firefox & Focus for iOS

joeyg replied:
I tested but only with the simple domains that are in the app now. I'll fix it to encode the URL and re-test

sashei commented on lockbox-ios/Store/UserInfoStore.swift:
```diff
@@ -5,6 +5,7 @@
 import Foundation
 import RxSwift
 import RxCocoa
+import UIKit
```
None of the other `Store`s import UIKit. Think carefully about the need to import this framework here; if I were doing this implementation, I would 1) write a helper extension to `UIApplication` and 2) put _all_ calls to outside services in an `ActionHandler`. Stores work internally to the application; `ActionHandler`s do the required outside calls. 

tbh, looking at this class again I'm not even sure that the `AutoLockSetting` enum belongs here anymore either...

sashei commented on lockbox-ios/Presenter/PreferredBrowserSettingPresenter.swift:
```diff
@@ -0,0 +1,73 @@
+            }
+
+            target.settingActionHandler.invoke(.preferredBrowser(browser: newPreferredBrowserValue))
+            }.asObserver()
```
nit: indentation, this line should be one tab back.

sashei commented on lockbox-ios/Presenter/PreferredBrowserSettingPresenter.swift:
```diff
@@ -0,0 +1,73 @@
+            }
+            .map { setting -> [CheckmarkSettingCellConfiguration] in
+                return self.initialSettings.map({ (cellConfiguration) -> CheckmarkSettingCellConfiguration in
+                    cellConfiguration.isChecked =
```
pick whether you want to leave these one line or break them -- you do two different things with your trinary operators here. you also have the option to disable swiftlint on a line-by-line basis if you choose using:
`// swiftlint:disable:this line_length`

sashei commented on lockbox-ios/View/PreferredBrowserSettingView.swift:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
```
we are `fatalError`ing everywhere else

sashei commented on lockbox-ios/View/SettingsView.swift:
```diff
@@ -167,6 +167,7 @@ class SwitchSettingCellConfiguration: SettingCellConfiguration {
 class CheckmarkSettingCellConfiguration: SettingCellConfiguration {
     var isChecked: Bool = false
     var valueWhenChecked: Any?
+    var enabled: Bool = true
```
could this be a property on all settings cells? it's possible that we will want to disable / enable other settings in the future.

sashei commented on lockbox-ios/Presenter/PreferredBrowserSettingPresenter.swift:
```diff
@@ -0,0 +1,73 @@
+}
+
+class PreferredBrowserSettingPresenter {
+    private var view: PreferredBrowserSettingViewProtocol
```
The reference to the `view` should always be `weak` to avoid retain cycles.

joeyg commented:
@sashei I updated the PR and moved the `UserInfoStore` code to a new handler. Let me know if I am going down the right path before I create the spec files.

sashei commented on lockbox-ios/Action/ExternalLinkAction.swift:
```diff
@@ -0,0 +1,99 @@
+    case Focus
+    case Safari
+
+    static var defaultValue: PreferredBrowserSetting {
```
Could this default live in the `Constants.swift` file with the other default settings?

sashei commented on lockbox-ios/Action/ExternalLinkAction.swift:
```diff
@@ -0,0 +1,99 @@
+
+        switch self {
+        case .Safari:
+            urlToOpen = URL(string: "http://\(url)")
```
why are you adding `http://`? most of the URLs are fully formed in the data models.

joeyg replied:
Ahh ok - I was going off the data in `populateTestData()`.

linuxwolf replied:
Let's assume the URLs will be fully-formed.  If we're really worried about it ... let's _not_ use "http://" when loading a likely login page

sashei commented on lockbox-ios/Action/ExternalLinkAction.swift:
```diff
@@ -0,0 +1,99 @@
+        case .Safari:
+            urlToOpen = URL(string: "http://\(url)")
+        case .Firefox:
+            urlToOpen = URL(string: "firefox://open-url?url=http://\(encodedString)")
```
^^ same question for firefox & focus

sashei commented:
@joeyg this approach is fine; feel free to move ahead with specs.

joeyg commented:
@devinreams no problem. I'll merge those in with my spec's.

devinreams commented:
@mozilla-lockbox/ux I'm adding you as a reviewer here. A few things I noticed quickly that you may want to compile into a list:

1. browser names: should we do full names for "Firefox Focus" and "Google Chrome"? (those are the names in the app store)
2. app URLs: Twitter app has claimed the twitter.com namespace, so tapping that link opens in the Twitter app itself, should we avoid that and force launching in the browser of choice?
3. link style: designs suggest a blue link-style text treatment, we should apply that text color here to indicate link-openability, right?

We can leave it up to @joeyg on whether this branch makes sense to address any requested changes in or if we should get this one "done" and spin up a separate issue for _anyone_ to pick up and address as follow-up work.

joeyg commented:
@devinreams re: browser names: Firefox Focus is called Klar in some markets (and uses `klar://` for deep linking). Not sure if that is something we want to support as well.

changecourse commented:
@devinreams addressing your points above.

1. Agreed, using the app store names for each browser makes sense to me
2. I think if that is the default functionality, we should support it (twitter.com opening the app instead of the site in the browser)
3. applying that style where? not sure I follow. 

Thanks for all of your work on this @joeyg, it's coming along nicely!

nickbrandt commented:
@changecourse I believe @devinreams was referencing the Entry Detail web address field (https://mozilla.invisionapp.com/share/BQEUSO0S2#/267560603_Entry_Detail), as this PR resolves #164 as well.

sashei commented on lockbox-ios/Action/ExternalLinkAction.swift:
```diff
@@ -0,0 +1,91 @@
+    }
+}
+
+class ExternalLinkActionHandler: ActionHandler {
```
I think you'll need to do the Cake pattern for dependency injection. Here's a pretty good blog post that uses `UIApplication` as an example: http://swiftyjimmy.com/mock-uiapplication-swift/

joeyg replied:
Thanks! this was helpful

sashei commented on lockbox-ios/Presenter/RootPresenter.swift:
```diff
@@ -200,6 +200,10 @@ class RootPresenter {
                 if !view.modalViewIs(AutoLockSettingView.self) {
                     view.pushSettingView(view: .autoLock)
                 }
+            case .preferredBrowser:
```
This isn't tested.

sashei commented on lockbox-ios/Action/ExternalLinkAction.swift:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    private func openUrl(string url: String, application: OpenUrlProtocol = UIApplication.shared) {
+        self.userDefaults.rx.observe(String.self, SettingKey.preferredBrowser.rawValue)
```
This can be another method in `UserDefaults+.swift`.

sashei commented on lockbox-ios/Action/ExternalLinkAction.swift:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    func canOpenBrowser(application: OpenUrlProtocol = UIApplication.shared) -> Bool {
+        if let url = getPreferredBrowserDeeplink(url: "https://mozilla.org") {
```
`self.getPreferred....`

sashei commented on lockbox-ios/Action/ExternalLinkAction.swift:
```diff
@@ -0,0 +1,91 @@
+    func openUrl(url: String,
+                 application: OpenUrlProtocol = UIApplication.shared,
+                 completion: ((Bool) -> Swift.Void)? = nil) {
+        if let urlToOpen = getPreferredBrowserDeeplink(url: url) {
```
`self.`

devinreams commented:
Yay! Code and tests are in place and reviewed. ✅ 

Just waiting on @mozilla-lockbox/product for a review which, since this PR has gone on a while, we can split off into follow-up issues to tackle.

devinreams commented:
> nickbrandt approved these changes
> sashei approved these changes

Yessss....

![](https://media.giphy.com/media/j3gsT2RsH9K0w/giphy.gif)

Thanks @joeyg 🙏

joeyg commented:
Thanks for everyone’s help! Glad to see this feature in the app :)

devinreams commented:
>full-width borders/separators on the options table

@nickbrandt Yeah that's going to be a consistent problem, definitely a great candidate for a `feature-ui` bug to highlight which views it's left on, it requires some hackery from what I attempted.

>disabling browser options that the user doesn't have installed

The treatment as implemented is to show a disabled/gray state seen here with "Google Chrome" on my phone:

![img_2623](https://user-images.githubusercontent.com/49511/38890434-a2a43460-423e-11e8-9780-c6de37524f27.jpg)
